packages/devtools/lib/src/framework/framework.dart

```diff
@@ -39,15 +45,93 @@ class Framework {
 
   Screen current;
 
+  Screen previous;
+
   StatusLine globalStatus;
   StatusLine pageStatus;
   StatusLine auxiliaryStatus;
   ActionsContainer globalActions;
   ConnectDialog connectDialog;
+  ImportMessage importMessage;
   AnalyticsOptInDialog analyticsDialog;
 
   final Map<Screen, CoreElement> _screenContents = {};
 
+  void _initDragDrop() {
+    window.addEventListener('dragenter', (e) => _onDragEnter(e), false);
+    window.addEventListener('dragover', (e) => _onDragOver(e), false);
+    window.addEventListener('dragleave', (e) => _onDragLeave(e), false);
+    window.addEventListener('drop', (e) => _onDrop(e), false);
+  }
+
+  void _onDragEnter(MouseEvent event) {
+    final Element dropTarget = event.target;
+    dropTarget.classes.add('over');
+  }
+
+  void _onDragOver(MouseEvent event) {
+    // This is necessary to allow us to drop.
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'move';
+  }
+
+  void _onDragLeave(MouseEvent event) {
+    final Element dropTarget = event.target;
+    dropTarget.classes.remove('over');
+  }
+
+  void _onDrop(MouseEvent event) async {
+    // Stop the browser from redirecting.
+    event.preventDefault();
+
+    final List<File> files = event.dataTransfer.files;
+    if (files.length > 1) {
+      toast('You cannot import more than one file.');
+      return;
+    }
+
+    final droppedFile = files.first;
+    if (droppedFile.type != 'application/json') {
+      toast('${droppedFile.type} is not a supported file type.\n Please import '
+          'a .json file.');
+      return;
+    }
+
+    final FileReader reader = FileReader();
+    reader.onLoad.listen((_) {
+      final Map<String, dynamic> import = jsonDecode(reader.result);
+      final source = import['source'];
+
+      // TODO(jacobr): add the inspector handling case here once the inspector
+      // can be exported.
+      switch (source) {
+        case timelineScreenId:
+          TimelineScreen timelineScreen = screens.firstWhere(
+            (screen) => screen.id == timelineScreenId,
+            orElse: () => null,
+          );
+          if (timelineScreen == null) {
+            addScreen(timelineScreen = TimelineScreen(disabled: false));
+          }
+          connectDialog.hide();
+          importMessage.hide();
+          navigateTo(timelineScreenId);
+          timelineScreen.loadFromImport(import);
+          break;
+        default:
+          toast('The imported file is from an unrecognized source page: '
```

Comment:
the user probably dragged in a json file that wasn't our file type. make the toast explain that instead?